plugins/embark-assistant/survey.cpp

```diff
@@ -1351,92 +1348,92 @@ void embark_assist::survey::survey_mid_level_tile(embark_assist::defs::geo_data
 
     for (uint8_t i = 1; i < 10; i++)
     {
-        if (survey_results->at(x).at(y).biome[i] != -1) {
-            biomes[survey_results->at(x).at(y).biome[i]] = true;
+        if (tile.biome[i] != -1) {
+            biomes[tile.biome[i]] = true;
         }
     }
     int count = 0;
     for (uint8_t i = 0; i <= ENUM_LAST_ITEM(biome_type); i++) {
         if (biomes[i]) count++;
     }
 
-    tile->biome_count = count;
+    tile.biome_count = count;
 
     for (uint8_t i = 0; i < 16; i++) {
-        tile->north_row[i].aquifer = mlt->at(i).at(0).aquifer;
-        tile->south_row[i].aquifer = mlt->at(i).at(15).aquifer;
-        tile->west_column[i].aquifer = mlt->at(0).at(i).aquifer;
-        tile->east_column[i].aquifer = mlt->at(15).at(i).aquifer;
-
-        tile->north_row[i].clay= mlt->at(i).at(0).clay;
-        tile->south_row[i].clay = mlt->at(i).at(15).clay;
-        tile->west_column[i].clay = mlt->at(0).at(i).clay;
-        tile->east_column[i].clay = mlt->at(15).at(i).clay;
-
-        tile->north_row[i].sand = mlt->at(i).at(0).sand;
-        tile->south_row[i].sand = mlt->at(i).at(15).sand;
-        tile->west_column[i].sand = mlt->at(0).at(i).sand;
-        tile->east_column[i].sand = mlt->at(15).at(i).sand;
-
-        tile->north_row[i].soil_depth = mlt->at(i).at(0).soil_depth;
-        tile->south_row[i].soil_depth = mlt->at(i).at(15).soil_depth;
-        tile->west_column[i].soil_depth = mlt->at(0).at(i).soil_depth;
-        tile->east_column[i].soil_depth = mlt->at(15).at(i).soil_depth;
-
-        tile->north_row[i].elevation = mlt->at(i).at(0).elevation;
-        tile->south_row[i].elevation = mlt->at(i).at(15).elevation;
-        tile->west_column[i].elevation = mlt->at(0).at(i).elevation;
-        tile->east_column[i].elevation = mlt->at(15).at(i).elevation;
-
-        tile->north_row[i].biome_offset = mlt->at(i).at(0).biome_offset;
-        tile->south_row[i].biome_offset = mlt->at(i).at(15).biome_offset;
-        tile->west_column[i].biome_offset = mlt->at(0).at(i).biome_offset;
-        tile->east_column[i].biome_offset = mlt->at(15).at(i).biome_offset;
-
-        tile->north_row[i].trees = mlt->at(i).at(0).trees;
-        tile->south_row[i].trees = mlt->at(i).at(15).trees;
-        tile->west_column[i].trees = mlt->at(0).at(i).trees;
-        tile->east_column[i].trees = mlt->at(15).at(i).trees;
+        tile.north_row[i].aquifer = mlt->at(i).at(0).aquifer;
```

Comment:
@PatrikLundell:
As I've said I refactored this in my index feature branch already, but thinking about it again I found my solution lacking.
My pet peeve is code duplication/repetition starting with the third time, as twice is often kind of okay.
And in this loop the same code is repeated 4 times.
So I measured what happened if we introduce a function to handle the copy process of the incursion data in comparison to the straight assignments.
Guess what :)
The following code is even faster than the original - not that it takes long anyways in the first place (13ms vs. 12,2ms for a large region on my system) - but the refactored version is easier to maintain as well.

```
void inline copy_incursion_values(embark_assist::defs::mid_level_tile_incursion_base &target, const embark_assist::defs::mid_level_tile_incursion_base &source) {
    target.aquifer = source.aquifer;
    target.clay = source.clay;
    target.sand = source.sand;
    target.soil_depth = source.soil_depth;
    target.elevation = source.elevation;
    target.biome_offset = source.biome_offset;
    target.trees = source.trees;
    target.savagery_level = source.savagery_level;
    target.evilness_level = source.evilness_level;
}

void refactored_incursion_loop(embark_assist::defs::region_tile_datum_new &tile, const embark_assist::defs::mid_level_tiles_new *mlt,
    const int8_t biome_corner[16][16], const int8_t biome_x[16][16], const int8_t biome_y[16][16]) {
    for (uint32_t i = 0; i < 257 * 257; i++) {
        for (uint8_t i = 0; i < 16; i++) {
            copy_incursion_values(tile.north_row[i], mlt->at(i).at(0));
            copy_incursion_values(tile.south_row[i], mlt->at(i).at(15));
            copy_incursion_values(tile.west_column[i], mlt->at(0).at(i));
            copy_incursion_values(tile.east_column[i], mlt->at(15).at(i));

            tile.north_corner_selection[i] = biome_corner[i][0];
            tile.west_corner_selection[i] = biome_corner[0][i];
            tile.north_row_biome_x[i] = biome_x[i][0];
            tile.west_column_biome_y[i] = biome_y[0][i];
        }
    }
}
```

Reply:
I agree introduction of the function is a better implementation that's easier to maintain. So, please, go ahead.

I wouldn't trust the performance measurement too much, though, as a compiler COULD optimize the code in a way that won't happen when used as intended (the corner assignments can be moved outside of the outer loop so it's done only 16 times). However, given the times we're talking about performance improvements aren't important.

Reply:
> I agree introduction of the function is a better implementation that's easier to maintain. So, please, go ahead.

Splendid - will do.

> I wouldn't trust the performance measurement too much, though, as a compiler COULD optimize the code in a way that won't happen when used as intended 

Very true.
> (the corner assignments can be moved outside of the outer loop so it's done only 16 times). 

Oh, the outer loop is just a helper construct to simulate all calls to `survey_mid_level_tile` in a large world. And in the feature branch I had to change around some things and ended up with the following lines, which remove the need for a loop completely:
```
std::copy(&details->edges.biome_x[0][0], &details->edges.biome_x[0][0] + 16 * 16, &tile.biome_x[0][0]);
std::copy(&details->edges.biome_y[0][0], &details->edges.biome_y[0][0] + 16 * 16, &tile.biome_y[0][0]);
std::copy(&details->edges.biome_corner[0][0], &details->edges.biome_corner[0][0] + 16 * 16, &tile.biome_corner[0][0]);
```

Reply:
I realized it was a simulation, but pointed at a possible loop hole in it, which a smart compiler could do.

I'm not familiar with std::copy (looking at a definition page), but from what I can understand you copy 256 elements from each of biome_x/y/corner rather than the 64 at the edges, which I guess means you store 256 rather 64 values as well..

Reply:
> I realized it was a simulation, but pointed at a possible loop hole in it, which a smart compiler could do.

Ah, now I get it - thanks for the clarification!

> which I guess means you store 256

Correct, to make the concurrent processing of internal and external incursions possible all `biome_*` data has to be available even if the "cursor" has moved to the next world tile already.